File: 哈希表/面试题 02.01.移除重复节点/移除重复节点.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeDuplicateNodes(head *ListNode) *ListNode {
	// 使用map来辅助解决问题
	// 同样的方法
	visted := make(map[int]bool)
	slownode, fastnode := head, head
	for fastnode != nil {
		_, ok := visted[fastnode.Val]
		if ok {
			slownode.Next = fastnode.Next
		} else {
			visted[fastnode.Val] = true
			slownode = fastnode
		}
		fastnode = fastnode.Next
	}
	return head
}

File: 哈希表/面试题 02.01.移除重复节点/移除重复节点.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
using namespace std;
// 上面的头文件与题解无光,只是为了在本地编译器里不报错

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// 上面与题解无光

class Solution {
public:
    ListNode* removeDuplicateNodes(ListNode* head) {
        unordered_set<int> myset;
        ListNode* prenode = nullptr;
        ListNode* fastnode = head;
        while (fastnode) {
            if (myset.count(fastnode->val)) {
                prenode->next = fastnode->next;
            } else {
                myset.insert(fastnode->val);
                prenode = fastnode;
            }
            fastnode = fastnode->next;
        }
        return head;
    }
};